File: ios/CylinderSeal/Services/UserPreferences.swift
import Combine
import Foundation

/// Lightweight preferences store. Uses `UserDefaults` because the only
/// sensitive material (the wallet private key) lives in the Keychain.
@MainActor
final class UserPreferences: ObservableObject {
    @Published private(set) var isOnboarded: Bool
    @Published private(set) var displayName: String
    @Published private(set) var phoneNumber: String?
    @Published private(set) var kycTier: KycTier
    @Published private(set) var superpeerHost: String
    @Published private(set) var superpeerPort: Int
    @Published private(set) var lastSyncAt: Date?

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.isOnboarded = defaults.bool(forKey: Keys.onboarded)
        self.displayName = defaults.string(forKey: Keys.displayName) ?? ""
        self.phoneNumber = defaults.string(forKey: Keys.phoneNumber)
        self.kycTier = KycTier(rawValue: defaults.string(forKey: Keys.kycTier) ?? "") ?? .anonymous
        self.superpeerHost = defaults.string(forKey: Keys.superpeerHost) ?? "sp-baghdad.cbi.iq"
        self.superpeerPort = defaults.integer(forKey: Keys.superpeerPort).nonZeroOr(50_051)
        self.lastSyncAt = defaults.object(forKey: Keys.lastSyncAt) as? Date
    }

    func completeOnboarding(displayName: String, phoneNumber: String?) {
        self.isOnboarded = true
        self.displayName = displayName
        self.phoneNumber = phoneNumber
        defaults.set(true, forKey: Keys.onboarded)
        defaults.set(displayName, forKey: Keys.displayName)
        if let phoneNumber { defaults.set(phoneNumber, forKey: Keys.phoneNumber) }
    }

    func setKycTier(_ tier: KycTier) {
        self.kycTier = tier
        defaults.set(tier.rawValue, forKey: Keys.kycTier)
    }

    func setSuperpeer(host: String, port: Int) {
        self.superpeerHost = host
        self.superpeerPort = port
        defaults.set(host, forKey: Keys.superpeerHost)
        defaults.set(port, forKey: Keys.superpeerPort)
    }

    func recordSync(at date: Date) {
        self.lastSyncAt = date
        defaults.set(date, forKey: Keys.lastSyncAt)
    }

    func reset() {
        for k in Keys.all { defaults.removeObject(forKey: k) }
        self.isOnboarded = false
        self.displayName = ""
        self.phoneNumber = nil
        self.kycTier = .anonymous
        self.lastSyncAt = nil
    }

    private enum Keys {
        static let onboarded = "cs.onboarded"
        static let displayName = "cs.displayName"
        static let phoneNumber = "cs.phoneNumber"
        static let kycTier = "cs.kycTier"
        static let superpeerHost = "cs.superpeerHost"
        static let superpeerPort = "cs.superpeerPort"
        static let lastSyncAt = "cs.lastSyncAt"

        static let all: [String] = [onboarded, displayName, phoneNumber, kycTier, superpeerHost, superpeerPort, lastSyncAt]
    }
}

private extension Int {
    func nonZeroOr(_ fallback: Int) -> Int {
        self == 0 ? fallback : self
    }
}

File: android/core/core-datastore/src/main/kotlin/com/modernecotech/cylinderseal/core/datastore/UserPreferences.kt
package com.modernecotech.cylinderseal.core.datastore

import android.content.Context
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

private val Context.dataStore by preferencesDataStore(name = "cs_prefs")

@Singleton
class UserPreferences @Inject constructor(@ApplicationContext private val ctx: Context) {

    private object Keys {
        val ONBOARDED = booleanPreferencesKey("onboarded")
        val DISPLAY_NAME = stringPreferencesKey("display_name")
        val PHONE_NUMBER = stringPreferencesKey("phone_number")
        val KYC_TIER = stringPreferencesKey("kyc_tier")
        val SUPERPEER_HOST = stringPreferencesKey("superpeer_host")
        val SUPERPEER_PORT = longPreferencesKey("superpeer_port")
        val LAST_SYNC_AT = longPreferencesKey("last_sync_at")
    }

    val isOnboarded: Flow<Boolean> = ctx.dataStore.data.map { it[Keys.ONBOARDED] ?: false }
    val displayName: Flow<String?> = ctx.dataStore.data.map { it[Keys.DISPLAY_NAME] }
    val phoneNumber: Flow<String?> = ctx.dataStore.data.map { it[Keys.PHONE_NUMBER] }
    val kycTier: Flow<String> = ctx.dataStore.data.map { it[Keys.KYC_TIER] ?: "ANONYMOUS" }
    val superpeerHost: Flow<String> =
        ctx.dataStore.data.map { it[Keys.SUPERPEER_HOST] ?: DEFAULT_HOST }
    val superpeerPort: Flow<Int> =
        ctx.dataStore.data.map { (it[Keys.SUPERPEER_PORT] ?: DEFAULT_PORT).toInt() }
    val lastSyncAt: Flow<Long> = ctx.dataStore.data.map { it[Keys.LAST_SYNC_AT] ?: 0L }

    suspend fun completeOnboarding(
        displayName: String,
        phoneNumber: String?,
    ) {
        ctx.dataStore.edit { p ->
            p[Keys.ONBOARDED] = true
            p[Keys.DISPLAY_NAME] = displayName
            phoneNumber?.let { p[Keys.PHONE_NUMBER] = it }
        }
    }

    suspend fun setKycTier(tier: String) {
        ctx.dataStore.edit { it[Keys.KYC_TIER] = tier }
    }

    suspend fun setSuperpeer(host: String, port: Int) {
        ctx.dataStore.edit {
            it[Keys.SUPERPEER_HOST] = host
            it[Keys.SUPERPEER_PORT] = port.toLong()
        }
    }

    suspend fun recordSync(nowMillis: Long) {
        ctx.dataStore.edit { it[Keys.LAST_SYNC_AT] = nowMillis }
    }

    /** Used by tests / settings "forget this device". */
    suspend fun reset() {
        ctx.dataStore.edit(Preferences::clear)
    }

    companion object {
        const val DEFAULT_HOST = "sp-baghdad.cbi.iq"
        const val DEFAULT_PORT = 50051L
    }
}
